libcaf_core/src/response_promise.cpp

```diff
@@ -67,32 +68,60 @@ bool response_promise::async() const {
   return id_.is_async();
 }
 
+local_actor* response_promise::self_ptr() const {
+  // TODO: We require that self_ was constructed by using a local_actor*. The
+  //       type erasure performed by strong_actor_ptr hides that fact. We
+  //       probably should use a different pointer type such as
+  //       intrusive_ptr<local_actor>. However, that would mean we would have to
+  //       include local_actor.hpp in response_promise.hpp or provide overloads
+  //       for intrusive_ptr_add_ref and intrusive_ptr_release.
+  auto self_baseptr = actor_cast<abstract_actor*>(self_);
+  return static_cast<local_actor*>(self_baseptr);
+}
 
 execution_unit* response_promise::context() {
-  return self_ == nullptr
-         ? nullptr
-         : static_cast<local_actor*>(actor_cast<abstract_actor*>(self_))
-           ->context();
+  return self_ == nullptr ? nullptr : self_ptr()->context();
 }
 
 void response_promise::deliver_impl(message msg) {
   CAF_LOG_TRACE(CAF_ARG(msg));
+  if (self_ == nullptr) {
+    CAF_LOG_DEBUG("drop response: invalid promise");
+    return;
+  }
+  auto self = self_ptr();
   if (!stages_.empty()) {
     auto next = std::move(stages_.back());
     stages_.pop_back();
-    next->enqueue(make_mailbox_element(std::move(source_), id_,
-                                       std::move(stages_), std::move(msg)),
-                  context());
+    detail::profiled_send(self, std::move(source_), next, id_,
+                          std::move(stages_), self->context(), std::move(msg));
+    self_.reset();
     return;
   }
   if (source_) {
-    source_->enqueue(std::move(self_), id_.response_id(),
-                     std::move(msg), context());
+    detail::profiled_send(self, self_, source_, id_.response_id(), no_stages,
```

Comment:
I think having `self` and `self_` here is a bit confusing. What differentiates them? (In `profiled_send.hpp` the second parameter is `sender`.)